File: 2-Android With Java/Day7/OFFLINE/Day5_Threads/app/src/main/java/com/example/day5_threads/ProductAdapter.java
package com.example.day5_threads;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.RequestOptions;

import java.util.List;

public class ProductAdapter extends RecyclerView.Adapter<ProductAdapter.ViewHolder>
{
    private Context context;
    private List<Product> productList;

    public ProductAdapter(Context context, List<Product> productList)
    {
        this.context = context;
        this.productList = productList;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType)
    {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.product, parent, false);
        return new ViewHolder(view);
    }


    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position)
    {
        Product product = productList.get(position);

        holder.title.setText(product.getTitle());
        holder.description.setText(product.getDescription());
        holder.price.setText(String.valueOf(product.getPrice()));
        holder.rating.setRating((float) product.getRating());
        holder.rating.setEnabled(false);

        // Load image with Glide, showing placeholder if offline
        RequestOptions requestOptions = new RequestOptions()
                .override(200, 200)
                .placeholder(R.drawable.ic_launcher_background)
                .error(R.drawable.ic_launcher_foreground);

        if (!NetworkUtil.isNetworkAvailable(context))
        {
            requestOptions = requestOptions.diskCacheStrategy(DiskCacheStrategy.ALL);
        }

        Glide.with(context)
                .load(product.getThumbnail())
                .apply(requestOptions)
                .into(holder.image);

        holder.constraintLayout.setOnClickListener(v -> {
            String message = product.getTitle();
            if (!NetworkUtil.isNetworkAvailable(context)) {
                message += " (offline)";
            }
            Toast.makeText(context, message, Toast.LENGTH_SHORT).show();
        });
    }

    @Override
    public int getItemCount()
    {
        return productList != null ? productList.size() : 0;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder
    {
        TextView title, description, price;
        ImageView image;
        RatingBar rating;
        ConstraintLayout constraintLayout;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.tittle);
            description = itemView.findViewById(R.id.description);
            price = itemView.findViewById(R.id.price);
            image = itemView.findViewById(R.id.imgv);
            rating = itemView.findViewById(R.id.ratingBar);
            constraintLayout = itemView.findViewById(R.id.rec);
        }
    }
}

File: 4-Android with Kotlin/Day5/app/src/main/java/com/example/day1/activity1/view/ProductAdapter.kt
package com.example.day1.activity1.view

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.example.day1.model.remote.Product
import com.example.day1.utils.ProductDiffUtill
import com.example.day1.R


class ProductAdapter(var myListener :(Product) -> Unit):ListAdapter<Product, ProductAdapter.ProuctViewHolder> (ProductDiffUtill())
{
    lateinit var context :Context

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProuctViewHolder {
        val inflater:LayoutInflater=LayoutInflater.from(parent.context)
        val view= inflater.inflate(R.layout.list_item ,parent , false)
        context= parent.context
        return ProuctViewHolder(view)
    }

    override fun onBindViewHolder(holder: ProuctViewHolder, position: Int) {
        val currentObj = getItem(position)
//        holder.Image.setImageResource(currentObj.thumbnail)
        holder.Name.text = currentObj.title
        holder.Layout.setOnClickListener {
            myListener.invoke(currentObj)
        }
        Glide.with(context)
            .load(currentObj.thumbnail)
            .apply(
                RequestOptions()
                    .override(200, 200)
                    .placeholder(R.drawable.ic_launcher_background)
                    .error(R.drawable.ic_launcher_foreground)
            )
            .into(holder.Image)
    }
    class ProuctViewHolder(private val item :View):RecyclerView.ViewHolder(item)
    {
        val Image :ImageView = item.findViewById(com.example.day1.R.id.productImage)
        val Name: TextView = item.findViewById(R.id.productName)
        val Layout : LinearLayout = item.findViewById(R.id.itemlayout)


    }
}